src/crimson/osd/osd_operations/scrub_event.h

```diff
@@ -0,0 +1,164 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <seastar/core/future.hh>
+
+#include <iostream>
+#include <variant>
+
+#include "crimson/common/type_helpers.h"
+#include "crimson/osd/osd_operations/client_request_common.h"
+#include "crimson/osd/osd_operations/common/pg_pipeline.h"
+#include "crimson/osd/osdmap_gate.h"
+#include "crimson/osd/pg_activation_blocker.h"
+#include "crimson/osd/pg_map.h"
+#include "messages/MOSDOp.h"
+#include "osd/PeeringState.h"
+#include "osd/osd_types.h"
+#include "osd/osd_types_fmt.h"
+#include "osd/scrubber_common.h"
+
+
+namespace crimson::osd {
+class ScrubEvent;
+}  // namespace crimson::osd
+
+namespace fmt {
+template <>
+struct formatter<crimson::osd::ScrubEvent>;
+}  // namespace fmt
+
+namespace crimson::osd {
+
+using namespace ::std::chrono;
+using namespace ::std::chrono_literals;
+
+class OSD;
+class ShardServices;
+class PG;
+
+
+class ScrubEvent : public PhasedOperationT<ScrubEvent> {
+ public:
+  static constexpr OperationTypeCode type = OperationTypeCode::scrub_event;
+
+  template <typename T = void>
+  using interruptible_future = ::crimson::interruptible::
+    interruptible_future<::crimson::osd::IOInterruptCondition, T>;
+
+  using ScrubEventFwdFut = interruptible_future<> (ScrubPgIF::*)(epoch_t);
+  using ScrubEventFwdImm = void (ScrubPgIF::*)(epoch_t);
+  using ScrubEventFwd = std::variant<ScrubEventFwdFut, ScrubEventFwdImm>;
+
+
+  class PGPipeline : public CommonPGPipeline {
+    struct AwaitMap : OrderedExclusivePhaseT<AwaitMap> {
+      static constexpr auto type_name = "ScrubEvent::PGPipeline::await_map";
+    } await_map;
+
+    //  do we need a pipe phase to lock the PG against other
+    //  types of client operations?
+
+    // a local synchronier, to enable us to finish an operation after creating
+    // a new event.
+    struct LocalSync : OrderedExclusivePhaseT<LocalSync> {
+      static constexpr auto type_name = "ScrubEvent::PGPipeline::local_sync";
+    } local_sync;
+
+    struct SendReply : OrderedExclusivePhaseT<SendReply> {
+      static constexpr auto type_name = "ScrubEvent::PGPipeline::send_reply";
+    } send_reply;
+
+    friend class ScrubEvent;
+  };
+
+ private:
+  Ref<PG> pg;
+  ScrubEventFwd event_fwd_func;
+  Scrub::act_token_t act_token;
+
+  PipelineHandle handle;
+  static PGPipeline& pp(PG& pg);  // should this one be static?
+
+ public:
+  struct nullevent_tag_t {
+  };
+
+
+ private:
+  ShardServices& shard_services;
+  spg_t pgid;
+  epoch_t epoch_queued;
+  std::chrono::milliseconds delay{0s};
+
+  const spg_t get_pgid() const { return pgid; }
+
+  virtual void on_pg_absent();
+  // virtual ScrubEvent::interruptible_future<> complete_rctx(Ref<PG>);
+  virtual seastar::future<> complete_rctx_no_pg() { return seastar::now(); }
+  virtual seastar::future<Ref<PG>> get_pg() final;
```

Comment:
If there is `final` or `override`, `virtual` stops bringing any further information to a person who reads the code.